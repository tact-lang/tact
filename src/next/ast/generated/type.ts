/* eslint-disable @typescript-eslint/no-unnecessary-condition */
import type * as $c from "@/next/ast/common";
import type * as $ from "@/next/ast/type";

export type Signedness = $.Signedness;
export const allSignedness: readonly $.Signedness[] = ["signed", "unsigned"];
export type IFInt = $.IFInt;
export const IFInt = (
    sign: $.Signedness,
    width: number,
    loc: $c.Range,
): $.IFInt =>
    Object.freeze({
        kind: "FInt",
        sign,
        width,
        loc,
    });
export const isIFInt = ($value: IFInt) => $value.kind === "FInt";
export type VarIntWidth = $.VarIntWidth;
export const allVarIntWidth: readonly $.VarIntWidth[] = ["16", "32"];
export type IFVarInt = $.IFVarInt;
export const IFVarInt = (
    sign: $.Signedness,
    width: $.VarIntWidth,
    loc: $c.Range,
): $.IFVarInt =>
    Object.freeze({
        kind: "FVarInt",
        sign,
        width,
        loc,
    });
export const isIFVarInt = ($value: IFVarInt) => $value.kind === "FVarInt";
export type IntFormat = $.IntFormat;
export type TypeInt = $.TypeInt;
export const TypeInt = (format: $.IntFormat, loc: $c.Range): $.TypeInt =>
    Object.freeze({
        kind: "TyInt",
        format,
        loc,
    });
export const isTypeInt = ($value: TypeInt) => $value.kind === "TyInt";
export type SFBits = $.SFBits;
export const SFBits = (bits: number, loc: $c.Range): $.SFBits =>
    Object.freeze({
        kind: "SFBits",
        bits,
        loc,
    });
export const isSFBits = ($value: SFBits) => $value.kind === "SFBits";
export type SFRemaining = $.SFRemaining;
export const SFRemaining = (loc: $c.Range): $.SFRemaining =>
    Object.freeze({
        kind: "SFRemaining",
        loc,
    });
export const isSFRemaining = ($value: SFRemaining) =>
    $value.kind === "SFRemaining";
export type SFDefault = $.SFDefault;
export const SFDefault = (loc: $c.Range): $.SFDefault =>
    Object.freeze({
        kind: "SFDefault",
        loc,
    });
export const isSFDefault = ($value: SFDefault) => $value.kind === "SFDefault";
export type SliceFormat = $.SliceFormat;
export type TypeSlice = $.TypeSlice;
export const TypeSlice = (format: $.SliceFormat, loc: $c.Range): $.TypeSlice =>
    Object.freeze({
        kind: "TySlice",
        format,
        loc,
    });
export const isTypeSlice = ($value: TypeSlice) => $value.kind === "TySlice";
export type RemFormat = $.RemFormat;
export type TypeCell = $.TypeCell;
export const TypeCell = (format: $.RemFormat, loc: $c.Range): $.TypeCell =>
    Object.freeze({
        kind: "TyCell",
        format,
        loc,
    });
export const isTypeCell = ($value: TypeCell) => $value.kind === "TyCell";
export type TypeBuilder = $.TypeBuilder;
export const TypeBuilder = (
    format: $.RemFormat,
    loc: $c.Range,
): $.TypeBuilder =>
    Object.freeze({
        kind: "TyBuilder",
        format,
        loc,
    });
export const isTypeBuilder = ($value: TypeBuilder) =>
    $value.kind === "TyBuilder";
export type TypeUnit = $.TypeUnit;
export const TypeUnit = (loc: $c.Range): $.TypeUnit =>
    Object.freeze({
        kind: "unit_type",
        loc,
    });
export const isTypeUnit = ($value: TypeUnit) => $value.kind === "unit_type";
export type TypeTensor = $.TypeTensor;
export const TypeTensor = (
    typeArgs: readonly $.Type[],
    loc: $c.Range,
): $.TypeTensor =>
    Object.freeze({
        kind: "tensor_type",
        typeArgs,
        loc,
    });
export const isTypeTensor = ($value: TypeTensor) =>
    $value.kind === "tensor_type";
export type TypeTuple = $.TypeTuple;
export const TypeTuple = (
    typeArgs: readonly $.Type[],
    loc: $c.Range,
): $.TypeTuple =>
    Object.freeze({
        kind: "tuple_type",
        typeArgs,
        loc,
    });
export const isTypeTuple = ($value: TypeTuple) => $value.kind === "tuple_type";
export type TypeCons = $.TypeCons;
export const TypeCons = (
    name: $c.TypeId,
    typeArgs: readonly $.Type[],
    loc: $c.Range,
): $.TypeCons =>
    Object.freeze({
        kind: "cons_type",
        name,
        typeArgs,
        loc,
    });
export const isTypeCons = ($value: TypeCons) => $value.kind === "cons_type";
export type Type = $.Type;
export type TypeMap = $.TypeMap;
export const TypeMap = (key: $.Type, value: $.Type, loc: $c.Range): $.TypeMap =>
    Object.freeze({
        kind: "map_type",
        key,
        value,
        loc,
    });
export const isTypeMap = ($value: TypeMap) => $value.kind === "map_type";
