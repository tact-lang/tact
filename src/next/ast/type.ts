import type { Range, TypeId } from "@/next/ast/common";

export type Type =
    | TypeMap
    | TypeCons
    | TypeInt
    | TypeSlice
    | TypeCell
    | TypeBuilder
    | TypeTuple
    | TypeUnit
    | TypeTensor;

export type TypeCons = {
    readonly kind: "cons_type";
    readonly name: TypeId;
    readonly typeArgs: readonly Type[];
    readonly loc: Range;
};

export type TypeInt = {
    readonly kind: "TyInt";
    readonly format: IntFormat;
    readonly loc: Range;
};
export type IntFormat = IFInt | IFVarInt;
export type IFInt = {
    readonly kind: "FInt";
    readonly sign: Signedness;
    readonly width: number;
    readonly loc: Range;
};
export type IFVarInt = {
    readonly kind: "FVarInt";
    readonly sign: Signedness;
    readonly width: VarIntWidth;
    readonly loc: Range;
};
export type VarIntWidth = "16" | "32";
export type Signedness = "signed" | "unsigned";

export type TypeSlice = {
    readonly kind: "TySlice";
    readonly format: SliceFormat;
    readonly loc: Range;
};
export type SliceFormat = SFBits | SFRemaining | SFDefault;
export type SFBits = {
    readonly kind: "SFBits";
    readonly bits: number;
    readonly loc: Range;
};
export type SFRemaining = {
    readonly kind: "SFRemaining";
    readonly loc: Range;
};
export type SFDefault = {
    readonly kind: "SFDefault";
    readonly loc: Range;
};

export type TypeCell = {
    readonly kind: "TyCell";
    readonly format: RemFormat;
    readonly loc: Range;
};
export type TypeBuilder = {
    readonly kind: "TyBuilder";
    readonly format: RemFormat;
    readonly loc: Range;
};
export type RemFormat = SFRemaining | SFDefault;

export type TypeTuple = {
    readonly kind: "tuple_type";
    readonly typeArgs: readonly Type[];
    readonly loc: Range;
};

export type TypeUnit = {
    readonly kind: "unit_type";
    readonly loc: Range;
};

export type TypeTensor = {
    readonly kind: "tensor_type";
    readonly typeArgs: readonly Type[];
    readonly loc: Range;
};

export type TypeMap = {
    readonly kind: "map_type";
    readonly key: Type; // any type except tensor
    readonly value: Type; // any type except tensor
    readonly loc: Range;
};
