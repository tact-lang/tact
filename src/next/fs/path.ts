/**
 * Safe relative path
 */
export type RelativePath = {
    /**
     * Number of "../" in front of path
     */
    readonly stepsUp: number;
    /**
     * /-separated strings that go after optional ../
     */
    readonly segments: readonly string[];
};
