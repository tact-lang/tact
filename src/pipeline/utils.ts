import type { TypeDescription } from "../types/types";

/**
 * This function sorts contracts in topological order.
 * It also checks for cycles in the graph of dependencies and returns undefined if a cycle is found.
 * It works in O(N) time, where N is the number of contracts.
 * @param allContracts - list of all contracts.
 * @returns sorted list of contracts or undefined if a cycle is found
 */
export const topSortContracts = (
    allContracts: TypeDescription[],
): TypeDescription[] | undefined => {
    const visitingNow: Set<TypeDescription> = new Set();
    const visited: Set<TypeDescription> = new Set();
    const result: TypeDescription[] = [];
    const dfs = (contract: TypeDescription): boolean => {
        visitingNow.add(contract);
        for (const c of contract.dependsOn) {
            if (c.kind !== "contract") {
                // This check should also be false. As contracts should be filtered out before calling this function.
                continue;
            }
            if (visitingNow.has(c)) {
                return false;
            }
            if (!visited.has(c)) {
                if (!dfs(c)) {
                    return false;
                }
            }
        }
        visitingNow.delete(contract);
        visited.add(contract);
        result.push(contract);
        return true;
    };
    for (const contract of allContracts) {
        if (!visited.has(contract)) {
            if (!dfs(contract)) {
                return undefined;
            }
        }
    }
    return result;
};
