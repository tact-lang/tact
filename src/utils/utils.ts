import { crc16 } from "./crc16";

export function topologicalSort<T>(src: T[], references: (src: T) => T[]) {
    const result: T[] = [];
    const visited = new Set<T>();
    const visiting = new Set<T>();
    const visit = (src: T) => {
        if (visiting.has(src)) {
            throw Error('Cycle detected');
        }
        if (!visited.has(src)) {
            visiting.add(src);
            for (const r of references(src)) {
                visit(r);
            }
            visiting.delete(src);
            visited.add(src);
            result.push(src);
        }
    }
    for (const s of src) {
        visit(s);
    }
    return result;
}

export function deepFreeze<T>(obj: T) {
    const propNames = Object.getOwnPropertyNames(obj);
    for (const name of propNames) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const value = (obj as any)[name];
        if (value && typeof value === "object") {
            deepFreeze(value);
        }
    }
    return Object.freeze(obj);
}

export function getMethodId(name: string) {
    return (crc16(name) & 0xffff) | 0x10000;
}