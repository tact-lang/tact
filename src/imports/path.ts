import { throwInternalCompilerError } from "../error/errors";
import { repeat } from "../utils/array";

// Witness tag. Do not use, do not export.
const pathTag = Symbol("path");

/**
 * Safe relative path
 */
export type RelativePath = {
    /**
     * Number of "../" in front of path
     */
    readonly stepsUp: number;
    /**
     * /-separated strings that go after optional ../
     */
    readonly segments: readonly string[];
    /**
     * Proof that path was created by RelativePath constructor
     */
    readonly [pathTag]: true;
};

/**
 * Constructor for relative paths
 */
const RelativePath = (
    stepsUp: number,
    segments: readonly string[],
): RelativePath => {
    if (stepsUp < 0) {
        throwInternalCompilerError("Negative number of ../ in path");
    }
    const result: RelativePath = {
        [pathTag]: true,
        stepsUp,
        segments: Object.freeze(segments),
    };
    return Object.freeze(result);
};

/**
 * Convert raw string with relative POSIX path into RelativePath
 */
export const fromString = (raw: string): RelativePath => {
    return raw.split("/").map(parseSegment).reduce(appendPath, emptyPath);
};

/**
 * Convert RelativePath to string.
 */
export const asString = ({ stepsUp, segments }: RelativePath): string => {
    return [...repeat("..", stepsUp), ...segments].join("/");
};

/**
 * Empty path, equivalent to "."
 */
export const emptyPath = RelativePath(0, []);

/**
 * Combine two relative paths
 */
const appendPath = (left: RelativePath, right: RelativePath): RelativePath => {
    const delta = right.stepsUp - left.segments.length;
    return RelativePath(left.stepsUp + Math.max(0, delta), [
        ...left.segments.slice(0, Math.max(0, -delta)),
        ...right.segments,
    ]);
};

const parseSegment = (segment: string): RelativePath => {
    if (segment === "..") {
        return RelativePath(1, []);
    }
    if (segment === "." || segment === "") {
        return emptyPath;
    }
    return RelativePath(0, [segment]);
};
