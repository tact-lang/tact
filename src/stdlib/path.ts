import path from "path";
import { posixNormalize } from "../utils/filePath";

export const stdlibPath = posixNormalize(path.join(__dirname, "stdlib"));
