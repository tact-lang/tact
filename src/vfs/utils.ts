import { sha256_sync } from "@ton/crypto";
import path from "path";

/**
 * Ensures the resulting file name does not exceed the given maximum length.
 * If too long, trims the name and appends a short hash to avoid collisions.
 *
 * @param name - The base file name without extension.
 * @param ext - The file extension.
 * @param maxLen - Maximum allowed length for the full file name (default: 255).
 * @returns A safe file name within the specified length.
 */
export const makeSafeName = (
    name: string,
    ext: string,
    maxLen = 255,
): string => {
    const full = name + ext;

    if (full.length <= maxLen) {
        return full;
    }

    const hash = sha256_sync(Buffer.from(name)).toString("hex").slice(0, 8);
    const suffix = `_${hash}${ext}`;
    const maxNameLen = maxLen - suffix.length;
    const safeName = name.slice(0, maxNameLen);

    return `${safeName}${suffix}`;
};

/**
 * Returns the full extension of a file, including all parts after the first dot.
 * - "file.txt" => ".txt"
 * - "archive.tar.gz" => ".tar.gz"
 */
export const getFullExtension = (filename: string): string => {
    const base = path.basename(filename);
    const firstDotIndex = base.indexOf(".");
    return firstDotIndex !== -1 ? base.slice(firstDotIndex) : "";
};
