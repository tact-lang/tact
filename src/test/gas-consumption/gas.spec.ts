import type {
    TransactionComputeVm,
    TransactionDescriptionGeneric,
} from "@ton/core";
import { beginCell, toNano } from "@ton/core";
import type {
    BlockchainTransaction,
    SandboxContract,
    TreasuryContract,
} from "@ton/sandbox";
import { Blockchain } from "@ton/sandbox";
import "@ton/test-utils";
import { getUsedGas } from "../../benchmarks/utils/gas";
import type { Step } from "../utils/write-vm-log";
import { writeLog } from "../utils/write-vm-log";
import { join } from "path";
import { Functions } from "./contracts/output/functions_Functions";
import { Forward } from "./contracts/output/forward_Forward";
import { Sha256Small } from "./contracts/output/benchmark_sha256_small_Sha256Small";
import { Sha256Big } from "./contracts/output/benchmark_sha256_big_Sha256Big";
import { Sha256AsSlice } from "./contracts/output/benchmark_sha256_as_slice_Sha256AsSlice";
import { CellsCreation } from "./contracts/output/cells_CellsCreation";
import { Addresses } from "./contracts/output/address_Addresses";
import { CodeOfVsInitOf } from "./contracts/output/codeOf_CodeOfVsInitOf";
import { WithDeploy } from "./contracts/output/deploy_WithDeploy";
import { WithoutDeploy } from "./contracts/output/deploy_WithoutDeploy";

function measureGas(txs: BlockchainTransaction[]): number {
    return Number(
        (
            (txs[1]!.description as TransactionDescriptionGeneric)
                .computePhase as TransactionComputeVm
        ).gasUsed,
    );
}

describe("benchmarks", () => {
    let blockchain: Blockchain;
    let treasure: SandboxContract<TreasuryContract>;

    let step: Step;

    beforeAll(async () => {
        blockchain = await Blockchain.create();

        step = writeLog({
            path: join(__dirname, "output", "log.yaml"),
            blockchain,
        });
    });

    beforeEach(async () => {
        treasure = await blockchain.treasury("benchmarks");
    });

    it("benchmark functions", async () => {
        const functions = blockchain.openContract(await Functions.fromInit());

        const sendResult = await step("benchmark functions", () =>
            functions.send(
                treasure.getSender(),
                { value: toNano(1) },
                { $$type: "Add", value: 10n },
            ),
        );
        const gasUsed = measureGas(sendResult.transactions);

        expect(gasUsed).toMatchSnapshot("gas used");

        // Verify code size
        const codeSize = functions.init!.code.toBoc().length;
        expect(codeSize).toMatchSnapshot("code size");
    });

    it("benchmark readFwdFee", async () => {
        const testContract = blockchain.openContract(await Forward.fromInit());
        const sendResult = await testContract.send(
            treasure.getSender(),
            { value: toNano(1) },
            {
                $$type: "TestGetFwdFee",
                any: beginCell()
                    .storeUint(0, 32)
                    .storeStringTail("This is test payload")
                    .asSlice(),
            },
        );
        const gasUsed = measureGas(sendResult.transactions);
        expect(gasUsed).toMatchSnapshot("gas used");
        const codeSize = testContract.init!.code.toBoc().length;
        expect(codeSize).toMatchSnapshot("code size");
    });

    async function hashStringSmall(
        sha256: SandboxContract<Sha256Small>,
        s: string,
    ): Promise<number> {
        const result = await sha256.send(
            treasure.getSender(),
            { value: toNano(1) },
            { $$type: "HashData", value: s },
        );

        return getUsedGas(result, "internal");
    }

    async function hashStringBig(
        sha256: SandboxContract<Sha256Big>,
        s: string,
    ): Promise<number> {
        const result = await sha256.send(
            treasure.getSender(),
            { value: toNano(1) },
            { $$type: "HashData", value: s },
        );

        return getUsedGas(result, "internal");
    }

    async function hashStringAsSLice(
        sha256: SandboxContract<Sha256AsSlice>,
        s: string,
    ): Promise<number> {
        const result = await sha256.send(
            treasure.getSender(),
            { value: toNano(1) },
            { $$type: "HashData", value: s },
        );

        return getUsedGas(result, "internal");
    }

    it("benchmark sha256", async () => {
        const sha256Small = blockchain.openContract(
            await Sha256Small.fromInit(),
        );
        const sha256Big = blockchain.openContract(await Sha256Big.fromInit());
        const sha256AsSlice = blockchain.openContract(
            await Sha256AsSlice.fromInit(),
        );

        await sha256Small.send(
            treasure.getSender(),
            { value: toNano(1) },
            null,
        );
        await sha256Big.send(treasure.getSender(), { value: toNano(1) }, null);
        await sha256AsSlice.send(
            treasure.getSender(),
            { value: toNano(1) },
            null,
        );

        await hashStringBig(sha256Big, "hello world");
        await hashStringSmall(sha256Small, "hello world");
        await hashStringAsSLice(sha256AsSlice, "hello world");

        expect(await hashStringBig(sha256Big, "hello world")).toMatchSnapshot(
            "gas hash string big",
        );
        expect(
            await hashStringSmall(sha256Small, "hello world"),
        ).toMatchSnapshot("gas hash string small");
        expect(
            await hashStringAsSLice(sha256AsSlice, "hello world"),
        ).toMatchSnapshot("gas hash string slice");

        expect(
            await hashStringBig(sha256Big, "hello world".repeat(5)),
        ).toMatchSnapshot("gas hash string big repeated");
        expect(
            await hashStringSmall(sha256Small, "hello world".repeat(5)),
        ).toMatchSnapshot("gas hash string small repeated");
        expect(
            await hashStringAsSLice(sha256AsSlice, "hello world".repeat(5)),
        ).toMatchSnapshot("gas hash string slice repeated");

        expect(
            await hashStringBig(sha256Big, "hello world".repeat(10)),
        ).toMatchSnapshot("gas hash string big repeated more");
        expect(
            await hashStringSmall(sha256Small, "hello world".repeat(10)),
        ).toMatchSnapshot("gas hash string small repeated more");
        expect(
            await hashStringAsSLice(sha256AsSlice, "hello world".repeat(10)),
        ).toMatchSnapshot("gas hash string slice repeated more");
    });

    it("benchmark cells creation", async () => {
        const testContract = blockchain.openContract(
            await CellsCreation.fromInit(),
        );
        await testContract.send(
            treasure.getSender(),
            { value: toNano(1) },
            null,
        );

        const gasUsed1 = (
            await blockchain.runGetMethod(testContract.address, "emptyCell")
        ).gasUsed;
        expect(gasUsed1).toMatchSnapshot("gas used emptyCell");

        const gasUsed2 = (
            await blockchain.runGetMethod(testContract.address, "emptySlice")
        ).gasUsed;
        expect(gasUsed2).toMatchSnapshot("gas used emptySlice");
    });

    it("benchmark contractAddressExt", async () => {
        const testContract = blockchain.openContract(
            await Addresses.fromInit(),
        );
        await testContract.send(
            treasure.getSender(),
            { value: toNano(1) },
            null,
        );
        const gasUsed = (
            await blockchain.runGetMethod(
                testContract.address,
                "contractAddressExt",
            )
        ).gasUsed;
        expect(gasUsed).toMatchSnapshot("gas used contractAddressExt");
    });

    it("benchmark codeOf vs initOf", async () => {
        const testContract = blockchain.openContract(
            await CodeOfVsInitOf.fromInit(),
        );
        await testContract.send(
            treasure.getSender(),
            { value: toNano(1) },
            null,
        );
        const gasUsed = (
            await blockchain.runGetMethod(testContract.address, "withCodeOf")
        ).gasUsed;
        expect(gasUsed).toMatchSnapshot("gas used withCodeOf");

        const gasUsed2 = (
            await blockchain.runGetMethod(testContract.address, "withInitOf")
        ).gasUsed;
        expect(gasUsed2).toMatchSnapshot("gas used withInitOf");
    });

    it("benchmark codeOf vs myCode()", async () => {
        const testContract = blockchain.openContract(
            await CodeOfVsInitOf.fromInit(),
        );
        await testContract.send(
            treasure.getSender(),
            { value: toNano(1) },
            null,
        );
        const gasUsed = (
            await blockchain.runGetMethod(testContract.address, "codeOfSelf")
        ).gasUsed;
        expect(gasUsed).toMatchSnapshot("gas used codeOf for current contract");

        const gasUsed2 = (
            await blockchain.runGetMethod(testContract.address, "myCode")
        ).gasUsed;
        expect(gasUsed2).toMatchSnapshot("gas used myCode");
    });

    it("benchmark deployable trait vs raw deploy", async () => {
        const withDeployTrait = blockchain.openContract(
            await WithDeploy.fromInit(),
        );
        const withoutDeploy = blockchain.openContract(
            await WithoutDeploy.fromInit(),
        );

        const deployResultTrait = await withDeployTrait.send(
            treasure.getSender(),
            { value: toNano(1) },
            {
                $$type: "Deploy",
                queryId: 1n,
            },
        );

        const deployRawResult = await withoutDeploy.send(
            treasure.getSender(),
            { value: toNano(1) },
            beginCell().endCell().beginParse(),
        );

        const gasUsed = measureGas(deployResultTrait.transactions);
        expect(gasUsed).toMatchSnapshot("gas used deploy trait");

        const gasUsedRaw = measureGas(deployRawResult.transactions);
        expect(gasUsedRaw).toMatchSnapshot("gas used raw deploy");
    });
});
