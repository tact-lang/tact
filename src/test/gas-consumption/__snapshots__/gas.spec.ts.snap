// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`benchmarks benchmark BaseTrait: code size 1`] = `352`;

exports[`benchmarks benchmark BaseTrait: gas used for ForwardMessage 1`] = `3586`;

exports[`benchmarks benchmark BaseTrait: gas used for NotifyMessage 1`] = `3393`;

exports[`benchmarks benchmark BaseTrait: gas used for ReplyMessage 1`] = `3460`;

exports[`benchmarks benchmark cashback: gas used cashback 1`] = `3337`;

exports[`benchmarks benchmark cells creation: gas used emptyCell 1`] = `750n`;

exports[`benchmarks benchmark cells creation: gas used emptySlice 1`] = `754n`;

exports[`benchmarks benchmark codeOf vs initOf: gas used withCodeOf 1`] = `850n`;

exports[`benchmarks benchmark codeOf vs initOf: gas used withInitOf 1`] = `1502n`;

exports[`benchmarks benchmark codeOf vs myCode(): gas used codeOf for current contract 1`] = `858n`;

exports[`benchmarks benchmark codeOf vs myCode(): gas used myCode 1`] = `858n`;

exports[`benchmarks benchmark contractAddressExt: gas used contractAddressExt 1`] = `2516n`;

exports[`benchmarks benchmark deployable trait vs raw deploy: gas used deploy trait 1`] = `3837`;

exports[`benchmarks benchmark deployable trait vs raw deploy: gas used raw deploy 1`] = `1505`;

exports[`benchmarks benchmark functions: code size 1`] = `170`;

exports[`benchmarks benchmark functions: gas used 1`] = `1979`;

exports[`benchmarks benchmark log: gas used log 1`] = `1884n`;

exports[`benchmarks benchmark readFwdFee: code size 1`] = `116`;

exports[`benchmarks benchmark readFwdFee: gas used 1`] = `1968`;

exports[`benchmarks benchmark sha256: gas hash string big 1`] = `2154`;

exports[`benchmarks benchmark sha256: gas hash string big repeated 1`] = `2155`;

exports[`benchmarks benchmark sha256: gas hash string big repeated more 1`] = `2157`;

exports[`benchmarks benchmark sha256: gas hash string slice 1`] = `2154`;

exports[`benchmarks benchmark sha256: gas hash string slice repeated 1`] = `2155`;

exports[`benchmarks benchmark sha256: gas hash string slice repeated more 1`] = `2157`;

exports[`benchmarks benchmark sha256: gas hash string small 1`] = `1975`;

exports[`benchmarks benchmark sha256: gas hash string small repeated 1`] = `1975`;

exports[`benchmarks benchmark sha256: gas hash string small repeated more 1`] = `1975`;

exports[`benchmarks benchmark sqrt via geom mean: code size 1`] = `265`;

exports[`benchmarks benchmark sqrt via geom mean: gas used for value 0 1`] = `2086`;

exports[`benchmarks benchmark sqrt via geom mean: gas used for value 100 1`] = `2978`;

exports[`benchmarks benchmark sqrt via geom mean: gas used for value 1000000000 1`] = `2978`;

exports[`benchmarks benchmark sqrt via geom mean: gas used for value 10000000000 1`] = `3185`;

exports[`benchmarks benchmark sqrt via geom mean: gas used for value 57896044618658097711785492504343953926634992332820282019728792003956564819968 1`] = `3806`;

exports[`benchmarks benchmark sqrt via geom mean: gas used for value 115792089237316195423570985008687907853269984665640564039457584007913129639935 1`] = `2564`;

exports[`benchmarks benchmark sqrt: code size 1`] = `265`;

exports[`benchmarks benchmark sqrt: gas used for value 0 1`] = `1916`;

exports[`benchmarks benchmark sqrt: gas used for value 100 1`] = `3305`;

exports[`benchmarks benchmark sqrt: gas used for value 1000000000 1`] = `3398`;

exports[`benchmarks benchmark sqrt: gas used for value 10000000000 1`] = `3398`;

exports[`benchmarks benchmark sqrt: gas used for value 57896044618658097711785492504343953926634992332820282019728792003956564819968 1`] = `3711`;

exports[`benchmarks benchmark sqrt: gas used for value 115792089237316195423570985008687907853269984665640564039457584007913129639935 1`] = `3711`;
